torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+        types: Tuple[Any],
+        args: Tuple[Any],
+        kwargs: Dict[str, Any],
+    ):
+        """
+        Lookup a cache entry for the given arguments. If none exists, dispatch
+        and cache the result (if the result is cache-able).
+        """
+        try:
+            key = self._cache_key(func, args, kwargs)
+        except _BypassDispatchCache as e:
+            FakeTensorMode.cache_bypasses[e.reason] += 1
+            return self._dispatch_impl(func, types, args, kwargs)
+
+        entry = FakeTensorMode.cache.get(key, None)
+        if entry is not None:
+            FakeTensorMode.cache_hits += 1
+            output = self._output_from_cache_entry(entry, func, args)
+        else:
+            FakeTensorMode.cache_misses += 1
+            output = self._dispatch_impl(func, types, args, kwargs)
+            try:
+                self._create_cache_entry(key, args, output)
+            except _BypassDispatchCache as e:
+                FakeTensorMode.cache_bypasses[e.reason] += 1
+
+        return output
+
+    def _cache_key(
+        self, func: OpOverload, args: Tuple[Any], kwargs: Dict[str, Any]
+    ) -> _DispatchCacheKey:
+        """
+        Create a cache key given the dispatch args. Raise _BypassDispatchCache
+        for any situation that precludes caching.
+        """
+        # Avoid caching for any ops that would require a more sophisticated
+        # caching implementation, e.g., data dependent ops or ops that modify
+        # the inputs.
+        if torch.Tag.data_dependent_output in func.tags:
+            raise _BypassDispatchCache("data dependent output")
+
+        if torch.Tag.dynamic_output_shape in func.tags:
+            raise _BypassDispatchCache("dynamic output shape")
+
+        if torch.Tag.inplace_view in func.tags:
+            raise _BypassDispatchCache("inplace view")
+
+        if func == aten._unsafe_view.default:
+            raise _BypassDispatchCache("unsafe view")
+
+        if func == aten.set_.source_Tensor:
+            raise _BypassDispatchCache("set source")
+
+        if func in self.lift_fns:
+            raise _BypassDispatchCache("lift")
+
+        # In order to handle storage aliasing, we need to establish the alias
+        # for any view op on a cache hit. But CompositeImplicitAutograd ops may
+        # or may not alias the input, so just punt on caching these.
+        if func.is_view and torch._C._dispatch_has_kernel_for_dispatch_key(
+            func.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        ):
+            raise _BypassDispatchCache("CompositeImplicitAutograd")
+
+        key_values = (
+            func,
+            # Translate any FakeTensor args to metadata.
+            self._translate_arg(args),
+            self._translate_arg(kwargs),
+            # Capture the default_dtype mode since that can affect the output tensor,
+            # e.g., when operating on constant float values.
+            torch.get_default_dtype(),
```

Comment:
good catch.. wonder if there's anything else we need to cache on globally 🤔 amp and grad_enabled should already be handled through torch_dispatch

Reply:
AFAICT, the device is always available in the Tensor so we capture it as part of the TensorMetadata. Are there non-Tensor situations where the default device could affect the operation?

Reply:
IIUC this code will happily cache tensor creation, where there isn't necessarily a tensor to take the device from. So default device should be relevant.

Reply:
@peterbell10 beat me too it :) (matters for tensor creation)

Reply:
Looks like `torch.backends.cuda.preferred_linalg_library()` can also change the output strides of linalg ops:
https://github.com/pytorch/pytorch/blob/afe6d272c69ae5671ca0df978be8fff7e8e4ed4e/aten/src/ATen/native/BatchLinearAlgebra.cpp#L682C62-L684

Reply:
Ugh, how many things like this could be out there? For this one, would people recommend adding that to the list of items we use to compute the key or skip caching linalg ops? cc @eellison

Reply:
That is a good point @peterbell10. However, I'll note that it's probably pretty unlikely that dynamo guards on this currently. If someone were to change the setting, we'd end up in a hard error at runtime when inductor checks strides, before or after this pr.

cc @yanboliang, is there a list somewhere of the all the global torch settings we guard on ? Maybe we could factor that out so that the fake tenosr caching here shares the logic.

Reply:
https://github.com/pytorch/pytorch/pull/111299 is trying to added guard on dynamo config, but it has been reverted due to internal SEV. I think we can guard on global torch settings in a similar way.